File: api/flpc/fmt.py
"""Response container + compact number formatting.

Every tool returns a `Response`: a title, a few context lines (filters, units),
one or more small tables and optional notes. It renders either as compact
CSV-style text (MCP — cheapest for an LLM to read) or as JSON (REST).
"""

from __future__ import annotations

import csv
import io
import math
from dataclasses import dataclass, field


class UserError(ValueError):
    """Bad input from the caller; the message is meant to be shown to the LLM."""


@dataclass
class Table:
    columns: list[str]
    rows: list[list]
    title: str | None = None


@dataclass
class Response:
    title: str
    context: list[str] = field(default_factory=list)
    tables: list[Table] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)

    def add(self, columns, rows, title=None) -> Table:
        t = Table(list(columns), [list(r) for r in rows], title)
        self.tables.append(t)
        return t

    def to_text(self) -> str:
        out = io.StringIO()
        out.write(f"## {self.title}\n")
        for c in self.context:
            out.write(c + "\n")
        w = csv.writer(out, lineterminator="\n")
        for t in self.tables:
            if t.title:
                out.write(f"\n# {t.title}\n")
            elif len(self.tables) > 1 or self.context:
                out.write("\n")
            w.writerow(t.columns)
            for r in t.rows:
                w.writerow(["" if v is None else v for v in r])
        if self.notes:
            out.write("\nNotes:\n")
            for n in self.notes:
                out.write(f"- {n}\n")
        return out.getvalue().rstrip() + "\n"

    def to_json(self) -> dict:
        return {
            "title": self.title,
            "context": self.context,
            "tables": [{"title": t.title, "columns": t.columns, "rows": t.rows} for t in self.tables],
            "notes": self.notes,
        }


# --------------------------------------------------------------------------- #
# Numbers
# --------------------------------------------------------------------------- #

def num(v, sig: int = 4):
    """Round for display: >=1000 -> int, else `sig` significant figures."""
    if v is None:
        return None
    if isinstance(v, float) and (math.isnan(v) or math.isinf(v)):
        return None
    if isinstance(v, int):
        return v
    a = abs(v)
    if a >= 1000 or v == int(v):
        return int(round(v))
    digits = sig - 1 - int(math.floor(math.log10(a)))
    r = round(v, max(digits, 0))
    return 0 if r == 0 else r


def pct(v, decimals: int = 1):
    """Fraction -> percent number (0.1234 -> 12.3)."""
    if v is None:
        return None
    r = round(v * 100, decimals)
    return 0.0 if r == 0 else r


USD_SCALES = [(1e9, "$B"), (1e6, "$M"), (1e3, "$K")]


class Scale:
    """Display scaling for one metric/unit across a whole table."""

    def __init__(self, unit: str, values=(), raw: bool = False):
        self.unit = unit
        self.div, self.label = 1.0, {"usd": "$", "usd_each": "$", "pct": "%",
                                     "count": "count", "rate": "per 1k"}.get(unit, unit)
        if unit == "usd" and not raw:
            vals = [abs(v) for v in values if v is not None]
            m = max(vals) if vals else 0
            for d, lab in USD_SCALES:
                if m >= d:
                    self.div, self.label = d, lab
                    break

    def __call__(self, v):
        if v is None:
            return None
        if self.unit == "pct":
            return pct(v, 2)
        if self.unit == "count":
            return int(round(v)) if abs(v) >= 1 or v == 0 else num(v)
        return num(v / self.div)

File: supabase/functions/_shared/flpc/fmt.ts
// Result container + compact number formatting (port of api/flpc/fmt.py).
//
// Every tool returns a `Result`: a title, a few context lines (filters,
// units), one or more small tables and optional notes. It renders as compact
// CSV-style text (MCP: cheapest for an LLM to read) or as JSON (REST).
// Rounding follows Python's round() exactly, so hosted and local answers match.

export type Cell = string | number | null;

/** Bad input from the caller; the message is meant to be shown to the LLM. */
export class UserError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UserError";
  }
}

export interface Table {
  columns: string[];
  rows: Cell[][];
  title: string | null;
}

function csvField(v: Cell): string {
  const s = v === null || v === undefined ? "" : String(v);
  return /[",\r\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

export class Result {
  title: string;
  context: string[] = [];
  tables: Table[] = [];
  notes: string[] = [];

  constructor(title: string) {
    this.title = title;
  }

  add(columns: string[], rows: Cell[][], title: string | null = null): Table {
    const t: Table = { columns: [...columns], rows: rows.map((r) => [...r]), title };
    this.tables.push(t);
    return t;
  }

  toText(): string {
    let out = `## ${this.title}\n`;
    for (const c of this.context) out += c + "\n";
    for (const t of this.tables) {
      if (t.title) out += `\n# ${t.title}\n`;
      else if (this.tables.length > 1 || this.context.length) out += "\n";
      out += t.columns.map(csvField).join(",") + "\n";
      for (const r of t.rows) out += r.map(csvField).join(",") + "\n";
    }
    if (this.notes.length) {
      out += "\nNotes:\n";
      for (const n of this.notes) out += `- ${n}\n`;
    }
    return out.trimEnd() + "\n";
  }

  toJSON() {
    return {
      title: this.title,
      context: this.context,
      tables: this.tables.map((t) => ({ title: t.title, columns: t.columns, rows: t.rows })),
      notes: this.notes,
    };
  }
}

// --------------------------------------------------------------------------
// Python-compatible rounding
// --------------------------------------------------------------------------

/** Python's round(x, n) for n >= 0: exact decimal value, ties to even. */
export function pyRound(x: number, n = 0): number {
  if (!Number.isFinite(x) || Math.abs(x) >= 1e21) return x;
  const neg = x < 0;
  const s = Math.abs(x).toFixed(100); // exact decimal expansion of the double
  const dot = s.indexOf(".");
  const digits = s.slice(0, dot) + s.slice(dot + 1);
  const cut = dot + n;
  let kept = digits.slice(0, cut);
  const rest = digits.slice(cut);
  let up = false;
  if (rest[0] > "5") up = true;
  else if (rest[0] === "5") {
    up = /[1-9]/.test(rest.slice(1)) || Number(kept[kept.length - 1] ?? "0") % 2 === 1;
  }
  if (up) {
    const arr = kept.split("");
    let i = arr.length - 1;
    while (i >= 0) {
      if (arr[i] === "9") {
        arr[i] = "0";
        i--;
      } else {
        arr[i] = String(Number(arr[i]) + 1);
        break;
      }
    }
    kept = (i < 0 ? "1" : "") + arr.join("");
  }
  const intLen = kept.length - n;
  const str = n > 0 ? `${kept.slice(0, intLen) || "0"}.${kept.slice(intLen)}` : kept;
  const r = Number(str);
  return neg ? -r : r;
}

/** Round for display: >= 1000 -> integer, else `sig` significant figures. */
export function num(v: number | null | undefined, sig = 4): number | null {
  if (v === null || v === undefined || !Number.isFinite(v)) return null;
  const a = Math.abs(v);
  if (a >= 1000 || v === Math.trunc(v)) return pyRound(v, 0) || 0;
  const digits = sig - 1 - Math.floor(Math.log10(a));
  const r = pyRound(v, Math.max(digits, 0));
  return r === 0 ? 0 : r;
}

/** Fraction -> percent number (0.1234 -> 12.3). */
export function pct(v: number | null | undefined, decimals = 1): number | null {
  if (v === null || v === undefined) return null;
  const r = pyRound(v * 100, decimals);
  return r === 0 ? 0 : r;
}

const USD_SCALES: [number, string][] = [[1e9, "$B"], [1e6, "$M"], [1e3, "$K"]];
const UNIT_LABEL: Record<string, string> = { usd: "$", usd_each: "$", pct: "%", count: "count", rate: "per 1k" };

/** Display scaling for one metric/unit across a whole table. */
export class Scale {
  unit: string;
  div = 1;
  label: string;

  constructor(unit: string, values: (number | null | undefined)[] = [], raw = false) {
    this.unit = unit;
    this.label = UNIT_LABEL[unit] ?? unit;
    if (unit === "usd" && !raw) {
      let m = 0;
      for (const v of values) if (v !== null && v !== undefined && Math.abs(v) > m) m = Math.abs(v);
      for (const [d, lab] of USD_SCALES) {
        if (m >= d) {
          this.div = d;
          this.label = lab;
          break;
        }
      }
    }
  }

  fmt(v: number | null | undefined): number | null {
    if (v === null || v === undefined) return null;
    if (this.unit === "pct") return pct(v, 2);
    if (this.unit === "count") return Math.abs(v) >= 1 || v === 0 ? pyRound(v, 0) || 0 : num(v);
    return num(v / this.div);
  }
}
